src/mds/query.py

```diff
@@ -64,22 +72,388 @@ async def search_metadata(
+    user-supplied pattern at any index in the "_resource_paths" array. For
+    example, given the following metadata objects:
+
+    {
+        "0": {
+            "message": "hello",
+            "_uploader_id": "100",
+            "_resource_paths": [
+                "/programs/a",
+                "/programs/b"
+            ],
+            "pet": "dog"
+        },
+        "1": {
+            "message": "greetings",
+            "_uploader_id": "101",
+            "_resource_paths": [
+                "/open",
+                "/programs/c/projects/a"
+            ],
+            "pet": "ferret",
+            "sport": "soccer"
+        },
+        "2": {
+            "message": "morning",
+            "_uploader_id": "102",
+            "_resource_paths": [
+                "/programs/d",
+                "/programs/e"
+            ],
+            "counts": [42, 42, 42],
+            "pet": "ferret",
+            "sport": "soccer"
+        },
+        "3": {
+            "message": "evening",
+            "_uploader_id": "103",
+            "_resource_paths": [
+                "/programs/f/projects/a",
+                "/admin"
+            ],
+            "counts": [1, 3, 5],
+            "pet": "ferret",
+            "sport": "basketball"
+        }
+    }
+
+    how do we design a filtering interface that allows the user to get all
+    objects having an authz string matching the pattern
+    "/programs/%/projects/%" at any index in its "_resource_paths" array? (%
+    has been used as the wildcard so far because that's what Postgres uses as
+    the wildcard for LIKE) In this case, the "1" and "3" objects should be
+    returned.
+
+    The filter syntax that was arrived at ending up following the syntax
+    specified by a Node JS implementation
+    (https://www.npmjs.com/package/json-api#filtering) of the JSON:API
+    specification (https://jsonapi.org/).
+
+    The format for this syntax is filter=(field_name,operator,value), in which
+    the field_name is a json key without quotes, operator is one of :eq, :ne,
+    :gt, :gte, :lt, :lte, :like, :all, :any (see operators dict), and value is
+    a typed json value against which the operator is run.
+
+    Examples:
+
+        - GET /objects?filter=(message,:eq,"morning") returns "2"
+        - GET /objects?filter=(counts.1,:eq,3) returns "3"
+
+    Compound expressions are supported:
+
+        - GET /objects?filter=(_resource_paths,:any,(,:like,"/programs/%/projects/%"))
+          returns "1" and "3"
+        - GET /objects?filter=(counts,:all,(,:eq,42))
+          returns "2"
+
+    Boolean expressions are also supported:
+
+        - GET /objects?filter=(or,(_uploader_id,:eq,"101"),(_uploader_id,:eq,"102"))
+          returns "1" and "2"
+        - GET /objects?filter=(or,(and,(pet,:eq,"ferret"),(sport,:eq,"soccer")),(message,:eq,"hello"))
+          returns "0", "1", and "2"
+
+    """
+    limit = min(limit, 2000)
```

Comment:
why 2000? (note that the indexd max limit is 1024)

Reply:
I went with `2000` just because that's what `search_metadata` uses.